Using AWS Distro for OpenTelemetry in EKS to ingest metrics into AMP recipe

dieteradant opened the pull request:
Added the following recipe. 

RFC, Suggestions?

dieteradant committed:
Initial commit

dieteradant committed:
Spelling fixes

mhausenblas reviewed (changes requested):
LGTM, here are a few things to address please:

1.  In the first paragraph describe the full setup (EKS, etc) and link to the services used (AMP, AMG, etc.
2.  Only the first time we use the full form of a service or OSS offering, so replace all but the first mentioning of, for example, Amazon Managed Service for Prometheus with AMP
3.  Don't use link titles such as "source" or "file" but make it self-descriptive. For example rather than "using the following template file" write "using `cluster_config.yaml`".
4. Make sure all services and offerings are properly spelled out (AWS fargate -> Fargate)
5. Rather than spreading prerequisites across the recipe, have a section at the top called "Prerequisites" where you list all of them (`ekstcl` comes to mind).
6. Whenever screen shots make sense, for example for "Make sure to add "Amazon Managed Service for Prometheus" as a datasource" then consider adding one.
7. Minor nit: pay attention to details, like left off ":"
8. Where ever you want users to validate something, such as in `kubectl get pods -n adot-col` make sure to show them an example output.

Thanks again and I'm looking forward to the next iteration.

dieteradant committed:
Moved Prerequisites to a seperate top section.

dieteradant committed:
Cleaned up some of the service names and other remarks

dieteradant committed:
Cleaned up some more text

dieteradant committed:
* Added image for "Make sure to add "Amazon Managed Service for Prometheus" as a datasource"
* Fixed broken links
* Fixed broken markdown for replacement vars

dieteradant commented:
I have taken into account all of these remarks, let me know if I missed something.

dieteradant committed:
Fixed some more remarks from Michael

dieteradant committed:
corrected some more remarks

dieteradant commented:
I have also taken into account the following points:

 1. Make sure to be consistent and attention to details for example Amazon Managed Service for Prometheus (AMP) vs Amazon Managed Service for Grafana AMG
2. We will be setting up an Amazon Elastic Kubernetes Service (EKS) cluster and Amazon Elastic Container Registry (ECR) repository to support this recipe.
3. You still have stuff like AWS Managed Service for Prometheus (AMP) in the text, as I said, only first time full then AMP
4. “using the following template cluster_config.yaml.” -> using cluster_config.yaml
5. Break multi-line commands or code, like:
6. Don’t do link only like “AWS Distro for OpenTelemetry (ADOT) Collector Setup”. turn it into sentences
7. Break down the Application section into multiple sub-headings (build, push)
8. Make sure that every section/subheading starts with a sentence that explains what we gonna do in that section

dieteradant commented:
Still todo:

@todo: Verify complete recipe technically
@todo: Create a Grafana dashboard screenshot from the end solution. (Replcace placeholder)

mhausenblas commented:
Thanks @dieteradant! @sheetaljoshi can you please verify the setup (try it out the recipe and provide comments?).

sheetaljoshi commented:
@dieteradant, thanks for adding this recipe. I will review and provide updates early next week.

mhausenblas commented:
Great, thanks @sheetaljoshi and can you please make a screen shot of the resulting Grafana board so we can update it?

sheetaljoshi commented on docs/recipes/ec2-eks-metrics-go-adot-ampamg.md:
```diff
@@ -1,12 +1,314 @@
+### Setup an EKS cluster
+
+For this recipe we require a EKS cluster to be available.
+You can either use an existing EKS cluster or create one using [cluster_config.yaml](ec2-eks-metrics-go-adot-ampamg/cluster_config.yaml).
```
Can we please address this issue, cluster_config.yaml does't resolve and results in 404, relative path reference is incorrect.

mhausenblas replied:
@sheetaljoshi not sure what you mean. It's part of this PR. In order to test it out you need to check out this PR. For example, using the `gh` CLI the command would be `gh pr checkout 5`.

dieteradant replied:
@sheetaljoshi Is this still an oustanding issue? Let me know how I can replicate this.

sheetaljoshi replied:
http://127.0.0.1:8000/recipes/ec2-eks-metrics-go-adot-ampamg/ec2-eks-metrics-go-adot-ampamg/cluster_config.yaml results into 404. Can you please check.

sheetaljoshi commented on docs/recipes/ec2-eks-metrics-go-adot-ampamg.md:
```diff
@@ -1,12 +1,314 @@
+
+#### Setup IAM role permissions for scraping
+
+1. Create the IAM role for the service account by following the steps in [Set up service roles for the ingestion of metrics from Amazon EKS clusters](https://docs.aws.amazon.com/prometheus/latest/userguide/set-up-irsa1.html#set-up-irsa-ingest). The AWS Distro for Open Telemetry Collector will use this role when it scrapes and exports metrics
```
External link doesn't take us to set-up-irsa-ingest, please fix. https://docs.aws.amazon.com/prometheus/latest/userguide/set-up-irsa.html#set-up-irsa-ingest

sheetaljoshi replied:
It is not concern to this PR, but the doc https://docs.aws.amazon.com/prometheus/latest/userguide/set-up-irsa.html#set-up-irsa-ingest says to replace my_prometheus_namespace, it is is not very clear what value we should provide here, if namespace: prometheus from cluster-config.yaml, it would be great if we can mention.

sheetaljoshi commented:
Thanks for all the great work! Here are few observations and things to consider please.

- It would be great to add to PR comments, as to where or which section do we find the new recipe i.e. By Compute->EKS->EC2->Metrics . This helps reviewer to locate a link to recipe vs. working backwords from list of files changed. 
- This comment is only for observation, have we decided to refer to external links for install and setups for prerequisites vs. including the steps along with recipes? It is completely ok, if this is the pattern we would like to establish for other recipes too. Reduces maintaining challenges related to tools (aws, kubectl, eksctl etc ) required to operate recipes.

mhausenblas commented:
Thanks for all your feedback @sheetaljoshi and I hope @dieteradant can address this week so we can launch beginning of next?

> It would be great to add to PR comments, as to where or which section do we find the new recipe i.e. By Compute->EKS->EC2->Metrics

Right. So, while it's linked already from https://github.com/open-o11y/aws-o11y-good-practices/blob/main/docs/eks.md I agree that as a good practice, going forward, folks should also include the "root" mentions. In this case, since I did that myself it's a bit murky but valid point!

> This comment is only for observation, have we decided to refer to external links for install and setups for prerequisites vs. including the steps along with recipes? It is completely ok, if this is the pattern we would like to establish for other recipes too. Reduces maintaining challenges related to tools (aws, kubectl, eksctl etc ) required to operate recipes.

That's a wonderful point. I was thinking to create a base recipe per compute unit/engine (ECS on Fargate or EKS on EC2) that we can refer to.

sheetaljoshi commented on docs/recipes/ec2-eks-metrics-go-adot-ampamg.md:
```diff
@@ -1,12 +1,314 @@
+
+4. Choose the Trust relationships tab and choose Edit trust relationship.
+
+5. In the trust relationship policy JSON, replace aws-amp with adot-col and then choose Update Trust Policy. Your resulting trust policy should look like the following: 
```
The trust relationshop policy had prometheus vs aws-amp in the role I created.

sheetaljoshi commented on docs/recipes/ec2-eks-metrics-go-adot-ampamg.md:
```diff
@@ -1,12 +1,314 @@
+
+Depending on if your EKS cluster uses AWS Fargate or not we will use a different template file for the rest of this guide.
+
+The following template file uses a deamonset for it's configuration [prometheus-daemonset.yaml](ec2-eks-metrics-go-adot-ampamg/prometheus-daemonset.yaml).
```
Going back to the beginning of this recipe, it says we will provision cluster with fargate profile, do we have to include methods for non fargate profile, it is confusing for consumers of this recipe in the first glance to decide between the right yaml.

dieteradant replied:
We will stick with only the fargate cluster in this recipe and split this out to a separate recipe later to avoid confusion.

sheetaljoshi commented on docs/recipes/ec2-eks-metrics-go-adot-ampamg.md:
````diff
@@ -1,12 +1,314 @@
+#### Deploy the template to your cluster
+
+```
+$ kubectl apply -f prometheus-fargate.yaml
````
This command doesn't work completely. Pods are not scheduled. Going back to the beginning of this page, we ask users to create cluster using fargate profile or atleast the cluster-config.yaml does that for you. And the cluster-config.yaml creates fargate profile for namespace prometheus, kube-system. So either a profile creating command is missing from the step or cluster-config.yaml should be updated to include odot-col namespace.

dieteradant replied:
I ended up adding the adot-col to the cluster-config.yaml

sheetaljoshi commented on docs/recipes/ec2-eks-metrics-go-adot-ampamg.md:
````diff
@@ -1,12 +1,314 @@
+You can verify that the ADOT Collector has started with this command:
+
+```
+$ kubectl get pods -n adot-col
````
The namespace issue needs to be fixed if using fargate profile.

dieteradant replied:
This namespace should now be available by adding it to the cluster-config.yaml

sheetaljoshi commented on docs/recipes/ec2-eks-metrics-go-adot-ampamg.md:
````diff
@@ -1,12 +1,314 @@
+
+### Build
+Clone the following Git repository
+```
````
Please fix the commands not to include $, use correct syntax highlighter. Copy paste does't work.

dieteradant replied:
Done

sheetaljoshi commented on docs/recipes/ec2-eks-metrics-go-adot-ampamg.md:
````diff
@@ -1,12 +1,314 @@
+Build the container
+```
+$ cd ./aws-otel-community/sample-apps/prometheus
+$ docker build . -t prometheus-sample-app:latest
````
It is a good idea to add a note for "go mod fails proxy.golang.or i/o timeout" due security restrictions on laptops.  change Dockerfile 
# RUN GO111MODULE=on go mod download
RUN GOPROXY=direct GO111MODULE=on go mod download

sheetaljoshi commented on docs/recipes/ec2-eks-metrics-go-adot-ampamg.md:
````diff
@@ -1,12 +1,314 @@
+### Push
+Authenticate to your default registry:
+```
+aws ecr get-login-password --region region | docker login --username AWS --password-stdin aws_account_id.dkr.ecr.region.amazonaws.com
````
Can we add notes to remind users  to provide region and account in this command, or use shell or env variables for AWS_REGION and ACCOUNT_ID.

sheetaljoshi commented on docs/recipes/ec2-eks-metrics-go-adot-ampamg.md:
````diff
@@ -1,12 +1,314 @@
+
+Push container to the ECR repository
+```
+docker tag prometheus-sample-app:latest <aws_account_id>.dkr.ecr.eu-west-1.amazonaws.com/prometheus-sample-app:latest
````
Please use the shell or env for account_id and region, please remove region hard coded here eu-west-1. It is preferred we stick to consistency.

sheetaljoshi commented on docs/recipes/ec2-eks-metrics-go-adot-ampamg.md:
````diff
@@ -1,12 +1,314 @@
+
+Deploy the sample app to your cluster:
+```
+$ kubectl apply -f prometheus-sample-app.yaml
````
This command will have the same challenge as prometheus-fargate.yaml if fargate profile chosen, I think it is a better idea to work this out.

dieteradant replied:
Resolved this one by adding it to the cluster-config.yml in the beginning just like the adot-col.

sheetaljoshi reviewed (commented):
I think we need to fix the basic ekscluster type issue, may be we create two recipes for eks cluster and one fargate profile.

sheetaljoshi commented:
Overall steps are good, left some inline comments. Not ready to merge yet. 

- Some of the basic questions to ask related to this recipe, do we want to stick to EKS cluster with managed node groups or also EKS cluster with fargate profile. Depending on what we decide, this recipe might need a fix to adjust some of the commands. 
- Experience can be improved with command copy/paste. Use inline shell variables or environment variables for AWS_REGION and AWS_ACCOUNT_ID to avoid user to leave thinking why the command failed, or add a note before every command where some of these values to be included.

dieteradant committed:
Fixed commands to not include $ to enable easy copy paste

dieteradant committed:
Added a note on goproxy

dieteradant committed:
fixed set-up-irsa-ingest link

dieteradant committed:
Made region and account id shell variables

dieteradant committed:
Added adot-col to the fargate profile

dieteradant committed:
removed the non-fargate documentation, added the sample app namespace to the fargate cluster

dieteradant committed:
The trust relationshop policy added info for the word prometheus

mhausenblas commented:
Quick status: @dieteradant and I reviewed today and he'll revise. In addition, we're trying to figure out why the ADOT collector doesn't seem to work for metrics on EKS Fargate.

dieteradant committed:
Fixed some extra notes from Michael

dieteradant committed:
AMG getting started fix

dieteradant committed:
Fixed the cluster config

dieteradant committed:
Split Fargate from EC2

dieteradant commented:
I have addressed most of the comments send by @mhausenblas.
Changes have been made to the recipe to revert back to EC2. 
I did a test run, but I might have overlooked some things.

mhausenblas commented:
@dieteradant thanks a lot and as soon as the remaining open issues (stuff at the bottom like "-> Put more content here.") are addressed, we're good to merge!

dieteradant committed:
Worked out some remaining open issues.

dieteradant committed:
Added some best practices for creating grafana dashboards

dieteradant commented:
> @dieteradant thanks a lot and as soon as the remaining open issues (stuff at the bottom like "-> Put more content here.") are addressed, we're good to merge!

I have worked out these remaining issues.

mhausenblas commented:
Super, thanks @dieteradant … the AMG screen shot as well?

dieteradant commented:
I don't have a nice AMG screenshot in the context of this recipe. 
I can either remove the placeholder or add it later.

mhausenblas commented:
@dieteradant roger that, no problem let's leave it for now and I fix it when I move the repo into its target GitHub org. Thanks for all your efforts here, much appreciated!
